File: lib/src/slug.dart
import 'package:unorm_dart/unorm_dart.dart' as unicode;

/// Identifies the input that made a namespaced key invalid.
enum NamespacedKeyField { namespace, key }

/// Validation detail returned by [namespacedKey].
final class NamespacedKeyValidationError {
  const NamespacedKeyValidationError({
    required this.field,
    required this.reason,
  });

  final NamespacedKeyField field;
  final String reason;

  @override
  bool operator ==(Object other) =>
      other is NamespacedKeyValidationError &&
      other.field == field &&
      other.reason == reason;

  @override
  int get hashCode => Object.hash(field, reason);

  @override
  String toString() => '${field.name} $reason';
}

/// Non-throwing result of constructing a namespaced key.
sealed class NamespacedKeyResult {
  const NamespacedKeyResult();

  bool get isValid;

  R match<R>({
    required R Function(String value) valid,
    required R Function(NamespacedKeyValidationError error) invalid,
  });
}

/// A validated `namespace:key` value.
final class ValidNamespacedKey extends NamespacedKeyResult {
  const ValidNamespacedKey(this.value);

  final String value;

  @override
  bool get isValid => true;

  @override
  R match<R>({
    required R Function(String value) valid,
    required R Function(NamespacedKeyValidationError error) invalid,
  }) =>
      valid(value);

  @override
  bool operator ==(Object other) =>
      other is ValidNamespacedKey && other.value == value;

  @override
  int get hashCode => value.hashCode;
}

/// An invalid namespace or key component.
final class InvalidNamespacedKey extends NamespacedKeyResult {
  const InvalidNamespacedKey(this.error);

  final NamespacedKeyValidationError error;

  @override
  bool get isValid => false;

  @override
  R match<R>({
    required R Function(String value) valid,
    required R Function(NamespacedKeyValidationError error) invalid,
  }) =>
      invalid(error);

  @override
  bool operator ==(Object other) =>
      other is InvalidNamespacedKey && other.error == error;

  @override
  int get hashCode => error.hashCode;
}

final RegExp _combiningMarks = RegExp(
  '[\u0300-\u036f\u1ab0-\u1aff\u1dc0-\u1dff\u20d0-\u20ff\ufe20-\ufe2f]',
);
final RegExp _notAsciiAlphaNumeric = RegExp('[^a-z0-9]+');
final RegExp _edgeHyphens = RegExp(r'^-+|-+$');

/// Normalizes [input] with NFKD and returns a lowercase ASCII kebab slug.
String slugify(String input) => unicode
    .nfkd(input)
    .replaceAll(_combiningMarks, '')
    .toLowerCase()
    .trim()
    .replaceAll(_notAsciiAlphaNumeric, '-')
    .replaceAll(_edgeHyphens, '');

/// Builds `namespace:key`, returning validation as data instead of throwing.
NamespacedKeyResult namespacedKey(String namespace, String key) {
  final String normalizedNamespace = slugify(namespace);
  if (normalizedNamespace.isEmpty) {
    return const InvalidNamespacedKey(
      NamespacedKeyValidationError(
        field: NamespacedKeyField.namespace,
        reason: 'must not be empty',
      ),
    );
  }

  final String normalizedKey = slugify(key);
  if (normalizedKey.isEmpty) {
    return const InvalidNamespacedKey(
      NamespacedKeyValidationError(
        field: NamespacedKeyField.key,
        reason: 'must not be empty',
      ),
    );
  }

  return ValidNamespacedKey('$normalizedNamespace:$normalizedKey');
}

File: tests/fixtures/api-baseline/lib/coreutils/slug.go
package coreutils

import (
	"strings"
	"unicode"

	"github.com/AtomiCloud/diene.go-errors-problems/lib/problem"
	"golang.org/x/text/unicode/norm"
)

// Slugify normalizes input with NFKD and returns a lowercase ASCII kebab slug.
func Slugify(input string) string {
	var builder strings.Builder
	separator := true
	for _, value := range norm.NFKD.String(input) {
		if unicode.Is(unicode.Mn, value) {
			continue
		}
		if value >= 'A' && value <= 'Z' {
			value += 'a' - 'A'
		}
		if (value >= 'a' && value <= 'z') || (value >= '0' && value <= '9') {
			builder.WriteRune(value)
			separator = false
			continue
		}
		if !separator && builder.Len() > 0 {
			builder.WriteByte('-')
			separator = true
		}
	}
	return strings.TrimSuffix(builder.String(), "-")
}

// NamespacedKey builds a normalized namespace:key value. Invalid components
// return a problem-typed validation error.
func NamespacedKey(namespace string, key string) (string, error) {
	normalizedNamespace := Slugify(namespace)
	if normalizedNamespace == "" {
		return "", namespacedKeyValidationError("namespace", "must not be empty")
	}
	normalizedKey := Slugify(key)
	if normalizedKey == "" {
		return "", namespacedKeyValidationError("key", "must not be empty")
	}
	return normalizedNamespace + ":" + normalizedKey, nil
}

func namespacedKeyValidationError(field string, message string) error {
	validation := problem.ValidationError()
	typeURI, _ := problem.TypeURI(problem.LocalErrorPortal(), validation.Version, validation.ID)
	return problem.NewError(problem.Problem{
		Type:        typeURI,
		Title:       validation.Title,
		Status:      validation.Status,
		Recoverable: validation.Recoverable,
		Data: map[string]any{"fields": []any{map[string]any{
			"path": field, "message": message,
		}}},
	})
}
